plugin/builtin/annotationstransformer/AnnotationsTransformer.go

```diff
@@ -24,7 +24,13 @@ func (p *plugin) Config(
 	_ *resmap.PluginHelpers, c []byte) (err error) {
 	p.Annotations = nil
 	p.FieldSpecs = nil
-	return yaml.Unmarshal(c, p)
+	if err = yaml.Unmarshal(c, p); err != nil {
+		return err
+	}
+	if p.FieldSpecs == nil {
+		p.FieldSpecs = annotations.DefaultFieldSpecs()
+	}
```

Comment:
I agree that we shouldn't swallow this error here; there are many reasons besides an empty `c` that `p.FieldSpecs` may be nil here.

Maybe an equivalent alternative that doesn't swallow errors for malformed values of `c` would be to check at the start of the function if `c` is empty, and in that case use the default fieldspecs?

Reply:
Maybe I'm being daft, but doesn't returning the error on unmarshal failure address that concern (i.e. the error is not swallowed, rendering the `p.FieldSpecs` clause unreachable in the event of failure). It's entirely likely that I'm misunderstanding, so please bear with me.

```go
	if err = yaml.Unmarshal(c, p); err != nil {
		return err
	}
```

Reply:
Ah, I think you're right, sorry I missed that. @yuwenma I don't think we are swallowing anything that wasn't previously swallowed. 

If there is a typo in `c`, e.g. `annotations-typo:` the transformer currently does nothing, silently.  
This PR will change it so that if there is a typo in `c`, the transformer will have the default fieldspecs. 

Neither is great, but I don't think this PR introduces a problem that wasn't already there, and I'm not sure how we would change it so I think this is fine.

Reply:
I believe [yaml.UnmarshalStrict](https://pkg.go.dev/sigs.k8s.io/yaml#UnmarshalStrict) disallows unknown fields, but perhaps that decision might be outside the scope of this particular behavioral change.

Reply:
The two examples are not for unmarshal error. Sorry I was distracted today. My main concern is the shifting behavior of transformers. Calling Config with passed-in `c` reads as "parsing the content of `c` and finding out the `annotations` pieces". So the valid return is either found c's content or nothing found (nil or error).  If the p.FieldSpecs is reset  to default, the `p.fieldspec` will be 
`[fieldspec{path: "metadata/annotations"},   fieldspec{path: "spec/template/metadata/annotations"},
fieldspec{path: "spec/template/metadata/annotations", kind: "Deployment"}]`

This is the swallowing info (or shifting behavior) I meant.

Reply:
@natasha41575 and I chatted about this today, and we're leaning towards wanting the new behaviour to be opt-in after all. Although we still agree that it seems very unlikely that using any builtin transformer with a nil fieldspec would make sense / have a useful result, it's also impossible to verify that. Knowing how important result stability is to our end users, that leads us to want to create a user-facing option for this. The use of fieldspecs in the configuration of our built in transformers is a convention, and we could enhance that convention by augmenting all the transformers that currently support it with an additional related field. 

```go
// Add the given annotations to the given field specifications.
type AnnotationsTransformerPlugin struct {
	Annotations map[string]string `json:"annotations,omitempty" yaml:"annotations,omitempty"`
	FieldSpecs  []types.FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
	FieldSpecStrategy FieldSpecStrategy `json:"fieldSpecStrategy" yaml:"fieldSpecStrategy"`
}
```

```yaml
apiVersion: builtin
kind: AnnotationsTransformer
metadata:
  name: not-important-to-example
annotations:
  app: myApp
  greeting/morning: a string with blanks
fieldSpecStrategy: merge | replace # defaults to replace for backwards compatibility
fieldSpecs:
- path: metadata/annotations
  create: true
```

With this solution:
- We immediately have a way to use the defaults as-is like you wanted, albeit requiring an extra LOC.
- We also immediately have a way to augment the defaults too.
- In Kustomization v1, we could start versioning the built-ins as you suggest. In the new version of these transformers, we could flip the default for this new field to "merge" (although who knows... by then, maybe the openapi-based approach will be ready instead).


WDYT?

Reply:
I appreciate the additional feedback and like the opt-in nature of your and @natasha41575's suggestion plus the stability it provides. And thank you @yuwenma for initiating this discussion. All the permutations make sense to me and I should be able to find some time to work on this tomorrow.

The one gap that I can conceive of with the merge behavior is perhaps the desire not to have it be exclusively additive, but instead out of the 10s of `commonLabels` field specs, perhaps one wishes to delete one or two via `$patch: delete` strategic merge semantics. Let me see what it will take to support that. I'm pretty sure examples abound within other parts of kustomize.

Reply:
The idea that the merge could work as a real SMP and support deletion is interesting for sure! Feel free to investigate it in a follow up though.